### ISSUE TITLE
Time Error

### ISSUE THREAD

**Issue by AravindRohit** [Soulter/hugging-chat-api#104]
```python
from hugchat import hugchat
from hugchat.login import Login
import time
from functools import wraps

COOKIE_PATH_DIR = "./cookies_snapshot"


def timer(func):
    @wraps(func)
    def wrapper(*args, **kwargs):
        start_time = time.time()
        result = func(*args, **kwargs)
        end_time = time.time()
        elapsed_time = end_time - start_time
        print(f"{func.__name__} took {elapsed_time:.2f} seconds to execute.")
        return result
    return wrapper


@timer
def chatbot_response(email, passwd, input_text):
    # Try to load cookies from the fixed directory
    try:
        sign = Login(email, None)
        cookies = sign.loadCookiesFromDir(COOKIE_PATH_DIR)
    except Exception as e:
        # If loading cookies fails, perform a new login
        sign = Login(email, passwd)
        cookies = sign.login()
        sign.saveCookiesToDir(COOKIE_PATH_DIR)

    # Create a ChatBot
    chatbot = hugchat.ChatBot(cookies=cookies.get_dict())

    # Create a new conversation
    conversation_id = chatbot.new_conversation()

    # Set the current conversation context
    chatbot.change_conversation(conversation_id)

    # Send the predefined input text to the chatbot
    response_dict = chatbot.query(
        text=input_text,
        temperature=0.2,
        max_new_tokens=1024,  # Experiment with a lower value
        top_k=10,  # Experiment with lower values
        top_p=0.7,  # Experiment with lower values
        use_cache=True,  # Enable response caching
        repetition_penalty=1.0,
        retry_count=2,  # Set the retry_count to a lower value, e.g., 2
        stream=False  # Disable streaming to get the entire response at once
    )

    # End the conversation (optional)
    chatbot.new_conversation()

    # Extract 'text' from the response dictionary
    if 'text' in response_dict:
        response_text = response_dict['text']
    else:
        response_text = "No 'text' field found in the response dictionary."

    # Ensure response is 2 to 3 sentences
    sentences = response_text.split('.')
    response_text = '. '.join(sentences[:3])

    # Ensure response ends with a full stop
    if not response_text.endswith('.'):
            response_text += '.'

    return response_text


# Replace these with your actual Hugging Face login credentials
email = "umadevi741401@gmail.com"
passwd = "Deviaravind@250723"
input_text = "what is recursive function in python?"

# Get the chatbot response
response = chatbot_response(email, passwd, input_text)

# Print the response
print(response)
```

This was my code it is taking 43 seconds to execute in latest version before version it is fast within 10 seconds how can we reduce it

**Comment by ogios** [comment]
the whole text maybe already done processing but stucked at the last `finalAnswer`, is it what's happening?

**Comment by RockChinQ** [comment]
@ogios is correct in some points, `hugchat` wait for last response before return your non-stream request now. But it should take same time in previous version, may I ask your version which can respond faster?

By the way, I tested your question `what is recursive function in python?` in non-stream mode with following code:

```python

from hugchat import hugchat
from hugchat.login import Login

import time

import logging

logging.basicConfig(level=logging.DEBUG)

print(time.time())

sign = Login("email", "pwd")
cookies = sign.login()

cookie_path_dir = "./cookies"
sign.saveCookiesToDir(cookie_path_dir)

print(time.time())

chatbot = hugchat.ChatBot(cookies=cookies.get_dict())

query_result = chatbot.chat("what is recursive function in python?")

print(query_result)

print(time.time())
```

And I got following output:

```
1696433043.451683
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): huggingface.co:443
DEBUG:urllib3.connectionpool:https://huggingface.co:443 "POST /login HTTP/1.1" 302 23
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): huggingface.co:443
DEBUG:urllib3.connectionpool:https://huggingface.co:443 "POST /chat/login HTTP/1.1" 200 488
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): huggingface.co:443
DEBUG:urllib3.connectionpool:https://huggingface.co:443 "GET /oauth/authorize?client_id=8f1a1d63-479b-46c8-84cb-521fe9f3222f&scope=openid%20profile&response_type=code&redirect_uri=https%3A%2F%2Fhuggingface.co%2Fchat%2Flogin%2Fcallback&state=eyJkYXRhIjp7ImV4cGlyYXRpb24iOjE2OTY0MzY2NDQyOTcsInJlZGlyZWN0VXJsIjoiaHR0cHM6Ly9odWdnaW5nZmFjZS5jby9jaGF0L2xvZ2luL2NhbGxiYWNrIn0sInNpZ25hdHVyZSI6IjUyM2Q2MTY3M2IwOTBiZjE0YWNhN2RjZTE5MWVlZTk1NGEwMWRiNGY3OTVjNmYwYTZjNTk2M2Y5Nzk2MTIxMDIifQ%3D%3D HTTP/1.1" 303 337
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): huggingface.co:443
DEBUG:urllib3.connectionpool:https://huggingface.co:443 "GET /chat/login/callback?code=hgknxhqPzxBaiGKK&state=eyJkYXRhIjp7ImV4cGlyYXRpb24iOjE2OTY0MzY2NDQyOTcsInJlZGlyZWN0VXJsIjoiaHR0cHM6Ly9odWdnaW5nZmFjZS5jby9jaGF0L2xvZ2luL2NhbGxiYWNrIn0sInNpZ25hdHVyZSI6IjUyM2Q2MTY3M2IwOTBiZjE0YWNhN2RjZTE5MWVlZTk1NGEwMWRiNGY3OTVjNmYwYTZjNTk2M2Y5Nzk2MTIxMDIifQ%3D%3D HTTP/1.1" 302 None
INFO:root:Cookie store path: ./cookies/1010553892@qq.com.json
1696433045.4028022
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): huggingface.co:443
DEBUG:urllib3.connectionpool:https://huggingface.co:443 "GET /chat HTTP/1.1" 308 None
DEBUG:urllib3.connectionpool:https://huggingface.co:443 "GET /chat/ HTTP/1.1" 200 None
DEBUG:urllib3.connectionpool:https://huggingface.co:443 "POST /chat/conversation HTTP/1.1" 200 None
DEBUG:root:{"conversationId":"651d839697974c89675f6445"}
DEBUG:urllib3.connectionpool:https://huggingface.co:443 "GET /chat/conversation/651d839697974c89675f6445/__data.json?x-sveltekit-invalidated=1_1 HTTP/1.1" 200 None
DEBUG:urllib3.connectionpool:https://huggingface.co:443 "POST /chat/conversation/651d839697974c89675f6445 HTTP/1.1" 200 None
DEBUG:urllib3.connectionpool:Resetting dropped connection: huggingface.co
DEBUG:urllib3.connectionpool:https://huggingface.co:443 "GET /chat/conversation/651d839697974c89675f6445/__data.json?x-sveltekit-invalidated=1_ HTTP/1.1" 200 None
A recursive function in Python is a function that calls itself, either directly or indirectly. In other words, a function solves a problem by breaking it down into smaller sub-problems, and then solving those sub-problems using the same function. This process continues until the solution to the original problem is found.

To define a recursive function in Python, you can use the `def` keyword followed by the name of the function and its parameters enclosed in parentheses. The function body contains one or more statements, including a call to the function itself. Here's an example:
```python
def factorial(n):
    if n == 0:
        return 1
    else:
        return n * factorial(n - 1)
```
In this example, the `factorial` function calculates the factorial of a given integer `n`. It does so by calling itself with a reduced value of `n - 1`, until it reaches the base case where `n` is equal to 0, at which point it returns the result directly.

Recursive functions have several characteristics:

1. Base case: There must be a base case that stops the recursion. In the above example, the base case is when `n` is equal to 0.
2. Recursion: The function must call itself, either directly or indirectly. In the above example, the function calls itself with `n - 1` as the argument.
3. Solution: The function must combine the solutions to the sub-problems to obtain the final solution. In the above example, the function multiplies the result of the recursive call by `n`.

Some tips for writing recursive functions in Python:

1. Use a clear and consistent naming convention for your variables and functions.
2. Make sure you have a well-defined base case that stops the recursion.
3. Use a debugger or print statements to trace the execution of your code and ensure it is working correctly.
4. Be mindful of the maximum recursion depth limit in Python, which is 1000 by default. You can increase this limit by setting the `sys.setrecursionlimit()` function.
5. Consider using iterative approaches instead of recursive ones when possible, especially for large inputs. Iterative approaches are generally faster and more memory-efficient.
1696433077.3071766
@RockChinQ ➜ /workspaces/hugging-chat-api/src (master) $

```

This request takes about `32s` which I think is normal. You can check your code or add debug output for each steps.

**Comment by ogios** [comment]
> This request takes about 32s which I think is normal.

i think it maybe a bit long for a normal message, except for web search.
did you use stream mode in this? try log.debug every response with the time.
i debugged this before, not until then i found that it will stuck at the final answer, but the thing is i can not determine if it's a aiohttp related problem or the request itself (i'm using aiohttp and they have this kind of issue on github)
how i fixed it is that i read response with small buffer and cut off connection once i captured it.

**Comment by AravindRohit** [comment]
> @ogios is correct in some points, `hugchat` wait for last response before return your non-stream request now. But it should take same time in previous version, may I ask your version which can respond faster?
> 
> By the way, I tested your question `what is recursive function in python?` in non-stream mode with following code:
> 
> ```python
> from hugchat import hugchat
> from hugchat.login import Login
> 
> import time
> 
> import logging
> 
> logging.basicConfig(level=logging.DEBUG)
> 
> print(time.time())
> 
> sign = Login("email", "pwd")
> cookies = sign.login()
> 
> cookie_path_dir = "./cookies"
> sign.saveCookiesToDir(cookie_path_dir)
> 
> print(time.time())
> 
> chatbot = hugchat.ChatBot(cookies=cookies.get_dict())
> 
> query_result = chatbot.chat("what is recursive function in python?")
> 
> print(query_result)
> 
> print(time.time())
> ```
> 
> And I got following output:
> 
> ```
> 1696433043.451683
> DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): huggingface.co:443
> DEBUG:urllib3.connectionpool:https://huggingface.co:443 "POST /login HTTP/1.1" 302 23
> DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): huggingface.co:443
> DEBUG:urllib3.connectionpool:https://huggingface.co:443 "POST /chat/login HTTP/1.1" 200 488
> DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): huggingface.co:443
> DEBUG:urllib3.connectionpool:https://huggingface.co:443 "GET /oauth/authorize?client_id=8f1a1d63-479b-46c8-84cb-521fe9f3222f&scope=openid%20profile&response_type=code&redirect_uri=https%3A%2F%2Fhuggingface.co%2Fchat%2Flogin%2Fcallback&state=eyJkYXRhIjp7ImV4cGlyYXRpb24iOjE2OTY0MzY2NDQyOTcsInJlZGlyZWN0VXJsIjoiaHR0cHM6Ly9odWdnaW5nZmFjZS5jby9jaGF0L2xvZ2luL2NhbGxiYWNrIn0sInNpZ25hdHVyZSI6IjUyM2Q2MTY3M2IwOTBiZjE0YWNhN2RjZTE5MWVlZTk1NGEwMWRiNGY3OTVjNmYwYTZjNTk2M2Y5Nzk2MTIxMDIifQ%3D%3D HTTP/1.1" 303 337
> DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): huggingface.co:443
> DEBUG:urllib3.connectionpool:https://huggingface.co:443 "GET /chat/login/callback?code=hgknxhqPzxBaiGKK&state=eyJkYXRhIjp7ImV4cGlyYXRpb24iOjE2OTY0MzY2NDQyOTcsInJlZGlyZWN0VXJsIjoiaHR0cHM6Ly9odWdnaW5nZmFjZS5jby9jaGF0L2xvZ2luL2NhbGxiYWNrIn0sInNpZ25hdHVyZSI6IjUyM2Q2MTY3M2IwOTBiZjE0YWNhN2RjZTE5MWVlZTk1NGEwMWRiNGY3OTVjNmYwYTZjNTk2M2Y5Nzk2MTIxMDIifQ%3D%3D HTTP/1.1" 302 None
> INFO:root:Cookie store path: ./cookies/1010553892@qq.com.json
> 1696433045.4028022
> DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): huggingface.co:443
> DEBUG:urllib3.connectionpool:https://huggingface.co:443 "GET /chat HTTP/1.1" 308 None
> DEBUG:urllib3.connectionpool:https://huggingface.co:443 "GET /chat/ HTTP/1.1" 200 None
> DEBUG:urllib3.connectionpool:https://huggingface.co:443 "POST /chat/conversation HTTP/1.1" 200 None
> DEBUG:root:{"conversationId":"651d839697974c89675f6445"}
> DEBUG:urllib3.connectionpool:https://huggingface.co:443 "GET /chat/conversation/651d839697974c89675f6445/__data.json?x-sveltekit-invalidated=1_1 HTTP/1.1" 200 None
> DEBUG:urllib3.connectionpool:https://huggingface.co:443 "POST /chat/conversation/651d839697974c89675f6445 HTTP/1.1" 200 None
> DEBUG:urllib3.connectionpool:Resetting dropped connection: huggingface.co
> DEBUG:urllib3.connectionpool:https://huggingface.co:443 "GET /chat/conversation/651d839697974c89675f6445/__data.json?x-sveltekit-invalidated=1_ HTTP/1.1" 200 None
> A recursive function in Python is a function that calls itself, either directly or indirectly. In other words, a function solves a problem by breaking it down into smaller sub-problems, and then solving those sub-problems using the same function. This process continues until the solution to the original problem is found.
> 
> To define a recursive function in Python, you can use the `def` keyword followed by the name of the function and its parameters enclosed in parentheses. The function body contains one or more statements, including a call to the function itself. Here's an example:
>     ```python
>     def factorial(n):
>         if n == 0:
>             return 1
>         else:
>             return n * factorial(n - 1)
>     ```
> In this example, the `factorial` function calculates the factorial of a given integer `n`. It does so by calling itself with a reduced value of `n - 1`, until it reaches the base case where `n` is equal to 0, at which point it returns the result directly.
> 
> Recursive functions have several characteristics:
> 
> 1. Base case: There must be a base case that stops the recursion. In the above example, the base case is when `n` is equal to 0.
> 2. Recursion: The function must call itself, either directly or indirectly. In the above example, the function calls itself with `n - 1` as the argument.
> 3. Solution: The function must combine the solutions to the sub-problems to obtain the final solution. In the above example, the function multiplies the result of the recursive call by `n`.
> 
> Some tips for writing recursive functions in Python:
> 
> 1. Use a clear and consistent naming convention for your variables and functions.
> 2. Make sure you have a well-defined base case that stops the recursion.
> 3. Use a debugger or print statements to trace the execution of your code and ensure it is working correctly.
> 4. Be mindful of the maximum recursion depth limit in Python, which is 1000 by default. You can increase this limit by setting the `sys.setrecursionlimit()` function.
> 5. Consider using iterative approaches instead of recursive ones when possible, especially for large inputs. Iterative approaches are generally faster and more memory-efficient.
> 1696433077.3071766
> @RockChinQ ➜ /workspaces/hugging-chat-api/src (master) $
> ```
> 
> This request takes about `32s` which I think is normal. You can check your code or add debug output for each steps.
> 
> P.S. Stream mode may be better if you want to track the progress of generating.

My previous hugchat version is 0.2.7 it is fast there 10 mins now in 0.2.8 is taking 35s

**Comment by AravindRohit** [comment]
It is depends on the question if we ask recursive function taking 40s but some other questions giving less time

**Comment by AravindRohit** [comment]
How can we run in ubuntu?

**Comment by RockChinQ** [comment]
@AravindRohit Upgrade to latest (v0.3.1) please.

**Comment by RockChinQ** [comment]
Hi guys, I added these code after the line `409` of `hugchat.py` to track progress, and tested again with this case https://github.com/Soulter/hugging-chat-api/issues/104#issuecomment-1747131050 :

```python
import time
logging.debug(f"[{time.time()}] {obj}")
```

And here's the whole output which show there's no unexpected delay while generating response:

https://pastebin.mozilla.org/OL5ME5gL

**Comment by ogios** [comment]
it takes that much of time for a long message then.
@AravindRohit does it happen in short responses?

**Comment by AravindRohit** [comment]
> @AravindRohit Upgrade to latest (v0.3.1) please.

It was in latest versions only

**Comment by RockChinQ** [comment]
> > @AravindRohit Upgrade to latest (v0.3.1) please.
> 
> It was in latest versions only

It's unexpected, but you can try my steps in the last comment to check the progress of response generating.

**Comment by AravindRohit** [comment]
It is generating but giving slow in latest versions

**Comment by AravindRohit** [comment]
Can You post your full code for verification

**Comment by RockChinQ** [comment]
Here's the whole `hugchat.py`, you can directly replace the `hugchat.py` with this content:

https://pastebin.mozilla.org/6QAacGBs

Then you can try the code below (you may need to make sure the `hugchat` module this code imported is the replaced one):

```python
from hugchat import hugchat
from hugchat.login import Login

import time

import logging

logging.basicConfig(level=logging.DEBUG)

print(time.time())

sign = Login("email", "pwd")
cookies = sign.login()

cookie_path_dir = "./cookies"
sign.saveCookiesToDir(cookie_path_dir)

print(time.time())

chatbot = hugchat.ChatBot(cookies=cookies.get_dict())

query_result = chatbot.chat("what is recursive function in python?")

print(query_result)

print(time.time())
```

And you'll get the similar output as the content in my comment yeasterday.

**Comment by RockChinQ** [comment]
Also, your comment remind me that I should compare the performance between this version and previous version (0.2.x) , I'll check it for sure.

**Comment by AravindRohit** [comment]
In my previous version 0.2.5 it is coming fast for every questions but now i am using 0.3.1 it is coming slowly for every questions

**Comment by RockChinQ** [comment]
I've checkout `v0.2.5` and added logging codes, the whole code of `hugchat.py` can be found here: https://pastebin.mozilla.org/HzbKCqwk . Please noticed that this version of code may not fit the latest protocol and has some unexpected bugs, but it can almost finish our test for this issue.

The test case is still previous one, and here's the output: https://pastebin.mozilla.org/GkgHFTJW   

The login output is incomplete due to vscode's terminal buffer, but the progress of generating is complete. This operation takes about 69s.

According to this output, we can now infer that this may be a problem of HuggingChat or network problems.

**Comment by AravindRohit** [comment]
How can we run hugchat in ubuntu?

**Comment by ogios** [comment]
@AravindRohit Just install with pip or pip3, install with apt `apt install python3-pip`
if you're refering to the error like this:
```
transfer-server on  main via 🐹 v1.21.0 took 4m37s 
❯ pip install
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try 'pacman -S
    python-xyz', where xyz is the package you are trying to
    install.
    
    If you wish to install a non-Arch-packaged Python package,
    create a virtual environment using 'python -m venv path/to/venv'.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip.
    
    If you wish to install a non-Arch packaged Python application,
    it may be easiest to use 'pipx install xyz', which will manage a
    virtual environment for you. Make sure you have python-pipx
    installed via pacman.

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.
```
then maybe try `apt install python-<package name>` ?

**Comment by AravindRohit** [comment]
> @AravindRohit Just install hugchat using pip or pip3, install pip with apt `apt install python3-pip` if you're refering to the error like this:
> 
> ```
> transfer-server on  main via 🐹 v1.21.0 took 4m37s 
> ❯ pip install
> error: externally-managed-environment
> 
> × This environment is externally managed
> ╰─> To install Python packages system-wide, try 'pacman -S
>     python-xyz', where xyz is the package you are trying to
>     install.
>     
>     If you wish to install a non-Arch-packaged Python package,
>     create a virtual environment using 'python -m venv path/to/venv'.
>     Then use path/to/venv/bin/python and path/to/venv/bin/pip.
>     
>     If you wish to install a non-Arch packaged Python application,
>     it may be easiest to use 'pipx install xyz', which will manage a
>     virtual environment for you. Make sure you have python-pipx
>     installed via pacman.
> 
> note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
> hint: See PEP 668 for the detailed specification.
> ```
> 
> then maybe try `apt install python-hugchat` ?

I need to  reduce the time for the responses

**Comment by Whitelisted1** [comment]
> 
> 
> 
> 
> > @AravindRohit Just install hugchat using pip or pip3, install pip with apt `apt install python3-pip` if you're refering to the error like this:
> 
> > 
> 
> > ```
> 
> > transfer-server on  main via 🐹 v1.21.0 took 4m37s 
> 
> > ❯ pip install
> 
> > error: externally-managed-environment
> 
> > 
> 
> > × This environment is externally managed
> 
> > ╰─> To install Python packages system-wide, try 'pacman -S
> 
> >     python-xyz', where xyz is the package you are trying to
> 
> >     install.
> 
> >     
> 
> >     If you wish to install a non-Arch-packaged Python package,
> 
> >     create a virtual environment using 'python -m venv path/to/venv'.
> 
> >     Then use path/to/venv/bin/python and path/to/venv/bin/pip.
> 
> >     
> 
> >     If you wish to install a non-Arch packaged Python application,
> 
> >     it may be easiest to use 'pipx install xyz', which will manage a
> 
> >     virtual environment for you. Make sure you have python-pipx
> 
> >     installed via pacman.
> 
> > 
> 
> > note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
> 
> > hint: See PEP 668 for the detailed specification.
> 
> > ```
> 
> > 
> 
> > then maybe try `apt install python-hugchat` ?
> 
> 
> 
> I need to  reduce the time for the responses

Try giving the same prompt on the [HuggingChat UI](https://huggingface.co/chat/). If the output takes the same amount of time with the same settings, then there is nothing we can do as HuggingFace's servers would be the bottleneck for generation speed.

**Comment by AravindRohit** [comment]
It is same in UI generation only

**Comment by ogios** [comment]
like what @Whitelisted1 said, you can test it in its webui
if it's a short message and it tooks that long to finish then maybe i have a lead.
but for long message, maybe it's just limited by the model's performance
